File: examples/basic.c
#include "ascon.h"
#include <stdio.h>
#include <string.h>

int main(void) {
    /* Demonstration only: provision a secret key and a fresh nonce in use. */
    const uint8_t key[ASCON_KEY_SIZE] = {0x42};
    const uint8_t nonce[ASCON_NONCE_SIZE] = {1};
    const uint8_t message[] = "Hello from C";
    const uint8_t ad[] = "example";
    uint8_t ciphertext[sizeof message - 1], plaintext[sizeof message - 1];
    uint8_t tag[ASCON_TAG_SIZE], digest[ASCON_HASH_SIZE], extended[48];

    int32_t status = ascon_aead128_encrypt(key, nonce, ad, sizeof ad - 1,
        message, sizeof message - 1, ciphertext, sizeof ciphertext, tag);
    if (status != ASCON_OK) return 1;
    status = ascon_aead128_decrypt(key, nonce, ad, sizeof ad - 1,
        ciphertext, sizeof ciphertext, tag, plaintext, sizeof plaintext);
    if (status != ASCON_OK) return 2;
    if (memcmp(plaintext, message, sizeof plaintext) != 0) return 3;
    if (ascon_hash256(message, sizeof message - 1, digest) != ASCON_OK) return 4;
    if (ascon_xof128(message, sizeof message - 1, extended, sizeof extended) != ASCON_OK) return 5;
    if (ascon_cxof128(message, sizeof message - 1, ad, sizeof ad - 1,
                      extended, sizeof extended) != ASCON_OK) return 6;
    puts("C encryption/decryption, Hash256, XOF128 and CXOF128 succeeded.");
    return 0;
}

File: examples/basic.rs
use ascon::{Aead128, Cxof128, Hash256};

fn main() -> Result<(), ascon::Error> {
    // Demonstration values only. Provision a secret key and a fresh nonce in use.
    let cipher = Aead128::new(&[0x42; 16]);
    let nonce = [0x01; 16];
    let mut message = *b"Hello from Rust";
    let tag = cipher.encrypt_in_place(&nonce, b"example", &mut message);
    println!("Ciphertext: {message:02x?}\nTag: {tag:02x?}");
    cipher.decrypt_in_place(&nonce, b"example", &mut message, &tag)?;
    assert_eq!(&message, b"Hello from Rust");

    let mut hash = Hash256::new();
    hash.update(b"Hello ");
    hash.update(b"from Rust");
    println!("Hash: {:02x?}", hash.finalize());

    let mut cxof = Cxof128::new(b"example protocol v1")?;
    cxof.update(&message);
    let mut reader = cxof.finalize();
    let mut output = [0; 48];
    reader.read(&mut output[..17]);
    reader.read(&mut output[17..]);
    println!("CXOF: {output:02x?}");
    Ok(())
}
